include/nbl/ext/ImGui/ImGui.h

```diff
@@ -1,76 +1,82 @@
 {
 class UI final : public core::IReferenceCounted
 {
 	public:
-		UI(core::smart_refctd_ptr<video::ILogicalDevice> device, int maxFramesInFlight, video::IGPURenderpass* renderpass, video::IGPUPipelineCache* pipelineCache, core::smart_refctd_ptr<ui::IWindow> window);
+		UI(core::smart_refctd_ptr<video::ILogicalDevice> _device, core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> _descriptorSetLayout, video::IGPURenderpass* renderpass, uint32_t subpassIx, video::IGPUPipelineCache* pipelineCache = nullptr, uint32_t mdiTotalByteSize = 1024u * 1024u * 2u /* 2Mb */);
 		~UI() override;
 
-		bool Render(nbl::video::IGPUCommandBuffer* commandBuffer, int frameIndex);
-		void Update(float deltaTimeInSec, float mousePosX, float mousePosY, size_t mouseEventsCount, ui::SMouseEvent const * mouseEvents); // TODO: Keyboard events
-		void BeginWindow(char const* windowName);
-		void EndWindow();
-		int Register(std::function<void()> const& listener);
-		bool UnRegister(int listenerId);
-		void SetNextItemWidth(float nextItemWidth);
-		void SetWindowSize(float width, float height);
-		void Text(char const* label, ...);
-		void InputFloat(char const* label, float* value);
-		void InputFloat2(char const* label, float* value);
-		void InputFloat3(char const* label, float* value);
-		void InputFloat4(char const* label, float* value);
-		void InputFloat3(char const* label, nbl::core::vector3df& value);
-		bool Combo(char const* label, int32_t* selectedItemIndex, char const** items, int32_t itemsCount);
-		bool Combo(const char* label, int* selectedItemIndex, std::vector<std::string>& values);
-		void SliderInt(char const* label, int* value, int minValue, int maxValue);
-		void SliderFloat(char const* label, float* value, float minValue, float maxValue);
-		void Checkbox(char const* label, bool* value);
-		void Spacing();
-		void Button(char const* label, std::function<void()> const& onPress);
-		void InputText(char const* label, std::string& outValue);
-		[[nodiscard]] bool HasFocus();		
-		[[nodiscard]] bool IsItemActive();
-		[[nodiscard]] bool TreeNode(char const* name);
-		void TreePop();
+		//! Nabla ImGUI backend reserves this index for font atlas, any attempt to hook user defined texture within the index will cause runtime error [TODO: could have a setter & getter to control the default & currently hooked font texture ID and init 0u by default]
+		_NBL_STATIC_INLINE_CONSTEXPR auto NBL_FONT_ATLAS_TEX_ID = 0u;
 
-	private:
+		//! update ImGUI internal state & cpu draw command lists, call it before this->render
+		bool update(const ui::IWindow* window, float deltaTimeInSec, const core::SRange<const nbl::ui::SMouseEvent> mouseEvents, const core::SRange<const nbl::ui::SKeyboardEvent> keyboardEvents);
+
+		//! updates mapped mdi buffer & records draw calls [TODO: maybe its a good idea to update mdi at the end of .update call instead of doing it here?]
+		bool render(nbl::video::SIntendedSubmitInfo& info, const nbl::video::IGPUDescriptorSet* const descriptorSet);
```

Comment:
answer to the TODO: no because you'd have to overflow submit in case of running out of space in MDI buffer

Reply:
resolved